lib/pure/os.nim

```diff
@@ -233,46 +233,79 @@ proc findExe*(exe: string, followSymlinks: bool = true;
-    for ext in extensions:
-      result = addFileExt(exe, ext)
-      if fileExists(result): return
-  when defined(posix):
-    if '/' in exe: checkCurrentDir()
-  else:
-    checkCurrentDir()
-  let path = getEnv("PATH")
-  for candidate in split(path, PathSep):
-    if candidate.len == 0: continue
-    when defined(windows):
-      var x = (if candidate[0] == '"' and candidate[^1] == '"':
-                substr(candidate, 1, candidate.len-2) else: candidate) /
-              exe
+  if exe.len != 0:
+    template checkCurrentDir() =
+     for ext in extensions:
+       var exeWithExt = addFileExt(exe, ext)
+       if fileExists(exeWithExt):
+         yield exeWithExt
+    when defined(posix):
+      if '/' in exe[1..^1]: checkCurrentDir()
     else:
-      var x = expandTilde(candidate) / exe
-    for ext in extensions:
-      var x = addFileExt(x, ext)
-      if fileExists(x):
-        when not (defined(windows) or defined(nintendoswitch)):
-          while followSymlinks: # doubles as if here
-            if x.symlinkExists:
-              var r = newString(maxSymlinkLen)
-              var len = readlink(x.cstring, r.cstring, maxSymlinkLen)
-              if len < 0:
-                raiseOSError(osLastError(), exe)
-              if len > maxSymlinkLen:
-                r = newString(len+1)
-                len = readlink(x.cstring, r.cstring, len)
-              setLen(r, len)
-              if isAbsolute(r):
-                x = r
-              else:
-                x = parentDir(x) / r
-            else:
-              break
-        return x
-  result = ""
+      checkCurrentDir()
+    let path = getEnv("PATH")
+    if path.len != 0:
+      for candidate in split(path, PathSep):
+        if candidate.len == 0: continue
+        when defined(windows):
+          var x = candidate.strip(chars = {'"'}) / exe
+        else:
+          var x = expandTilde(candidate) / exe
+        var checks = 0
+        for ext in extensions:
+          if checks >= maxChecksPerDirectory: break
+          var x = addFileExt(x, ext)
+          if fileExists(x):
+            checks += 1
+            when not (defined(windows) or defined(nintendoswitch)):
+              while followSymlinks: # combines while true & if followSymlinks
+                if x.symlinkExists:
+                  var r = newString(maxSymlinkLen)
+                  var len = readlink(x.cstring, r.cstring, maxSymlinkLen)
+                  if len < 0:
+                    raise newException(OSError, "Failed to follow symlink" & x)
+                  if len > maxSymlinkLen:
+                    r = newString(len + 1)
+                    len = readlink(x.cstring, r.cstring, len)
+                  setLen(r, len)
+                  if isAbsolute(r):
+                    x = r
+                  else:
+                    x = parentDir(x) / r
+                else:
+                  break
+            yield x
+
+proc findExe*(exe: string, followSymlinks: bool=true, maxChecks=1000,
```

Comment:
`maxChecks = 1000.Positive` IMHO.

(Positive is not perfect for advanced users, but still better than just plain int because it wont allow negative values)